// @uipath/robot/dist/models/jobPromise.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var uipathRobot_1 = require("../uipathRobot");
var constants_1 = require("../utils/constants");
/**
 * Job promise model used for resolution and/or rejection of Promise<JobResult>
 */
var JobPromise = /** @class */ (function () {
    /**
     * Default constructor
     * @param job Job to be wrapped with promise
     */
    function JobPromise(job) {
        var _this = this;
        this.job = job;
        /**
         * Method to attach job status event handlers.
         * @param eventHanlder Attaches callback which is invoked on job status change.
         */
        this.onStatus = function (eventHanlder) {
            _this.job.on(constants_1.Constants.STATUS, eventHanlder);
            return _this;
        };
        this.promise = uipathRobot_1.UiPathRobot.startJob(this.job);
    }
    JobPromise.prototype.then = function (onfulfilled, onrejected) {
        return this.promise.then(onfulfilled, onrejected);
    };
    JobPromise.prototype.catch = function (onrejected) {
        return this.promise.catch(onrejected);
    };
    JobPromise.prototype.finally = function (onfinally) {
        return this.promise.finally(onfinally);
    };
    return JobPromise;
}());
exports.JobPromise = JobPromise;
