// @uipath/robot/dist/utils/dispatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Dispatcher class to allow event handling on custom classes.
 */
var Dispatcher = /** @class */ (function () {
    function Dispatcher() {
        var _this = this;
        /**
         * Method to determine if a certain event has been registered.
         * @returns Boolean result
         */
        this.has = function (eventName) { return _this.dispatchHandlers.has(eventName); };
        this.dispatchHandlers = new Map();
    }
    /**
     * Event registration method.
     * @param eventName Name of the event to register.
     * @param handler callback function to call when the event occurs.
     */
    Dispatcher.prototype.on = function (eventName, handler) {
        var allHandlers = this.dispatchHandlers.get(eventName);
        if (allHandlers && this.dispatchHandlers.has(eventName)) {
            allHandlers.push(handler);
        }
        else {
            this.dispatchHandlers.set(eventName, [handler]);
        }
    };
    /**
     * Dispatch method to dispatch registered callbacks on the occuring event.
     * @param eventName Name of the event that has occured.
     * @param argument Arguments to pass to callback if any.
     */
    Dispatcher.prototype.dispatch = function (eventName, argument) {
        var _this = this;
        var allHandlers = this.dispatchHandlers.get(eventName);
        if (allHandlers && this.dispatchHandlers.has(eventName)) {
            allHandlers.forEach(function (handler) {
                handler.apply(_this, [argument]);
            });
        }
    };
    return Dispatcher;
}());
exports.Dispatcher = Dispatcher;
