// @uipath/robot/dist/requests/implementation/portRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../utils");
var constants_1 = require("../../utils/constants");
/**
 * Request class to get port of listener running on local user session.
 */
var PortRequest = /** @class */ (function () {
    /**
     * Default constructor
     * @param listenerPort Port of listener running on session 0.
     * @param consentCode User approved consent code if any.
     */
    function PortRequest(listenerPort, consentCode) {
        var _this = this;
        /**
        * Method that sends the request to get port of listener running on local user session.
        * @returns Promise of Response model type.
        */
        this.getPort = function () { return utils_1.HttpUtil.get(_this.url, _this.convertToPort); };
        /**
         * Mapper method used to convert string response to number.
         * @param response HttpResponse sent from listener.
         * @returns Port number.
         */
        this.convertToPort = function (response) {
            if (response) {
                var portResponse = JSON.parse(response);
                return portResponse.Port;
            }
            return 0;
        };
        /**
         * Method that sets the request endpoint to which request will be sent.
         * @param listenerEndpoint Endpoint to which request will be sent
         * @param consentCode randomly generated consent code.
         */
        this.setURL = function (listenerEndpoint, consentCode) { return utils_1.UrlUtil.appendConsentCode(listenerEndpoint + "/" + constants_1.Constants.GET_PORT, consentCode); };
        var listenerEndpoint = constants_1.Constants.DEFAULT_PORT_DISCOVERY_DOMAIN + ":" + listenerPort;
        this.url = this.setURL(listenerEndpoint, consentCode);
    }
    return PortRequest;
}());
exports.PortRequest = PortRequest;
